dbt/adapters/spark/impl.py

```diff
@@ -209,69 +259,79 @@ def find_table_information_separator(rows: List[dict]) -> int:
         return pos
 
     def get_columns_in_relation(self, relation: Relation) -> List[SparkColumn]:
-        cached_relations = self.cache.get_relations(relation.database, relation.schema)
-        cached_relation = next(
-            (
-                cached_relation
-                for cached_relation in cached_relations
-                if str(cached_relation) == str(relation)
-            ),
-            None,
-        )
+        # We shouldn't access columns from the cache, until we've implemented
+        # proper cache update or invalidation at the column level
+        # https://github.com/dbt-labs/dbt-spark/issues/431
+
+        # cached_relations = self.cache.get_relations(relation.database, relation.schema)
+        # cached_relation = next(
+        #     (
+        #         cached_relation
+        #         for cached_relation in cached_relations
+        #         if str(cached_relation) == str(relation)
+        #     ),
+        #     None,
+        # )
+
+        # For now, just always invalidate the cache
+        updated_relation = self._get_updated_relation(relation)
+        if updated_relation:
+            self.cache.add(updated_relation)
+            return updated_relation.columns
+        else:
+            return []
+
+    def _get_updated_relation(self, relation: BaseRelation) -> Optional[SparkRelation]:
+        metadata = None
         columns = []
-        if cached_relation and cached_relation.information:
-            columns = self.parse_columns_from_information(cached_relation)
-        if not columns:
-            # in open source delta 'show table extended' query output doesnt
-            # return relation's schema. if columns are empty from cache,
-            # use get_columns_in_relation spark macro
-            # which would execute 'describe extended tablename' query
-            try:
-                rows: List[agate.Row] = self.execute_macro(
-                    GET_COLUMNS_IN_RELATION_RAW_MACRO_NAME, kwargs={"relation": relation}
-                )
-                columns = self.parse_describe_extended(relation, rows)
-            except dbt.exceptions.RuntimeException as e:
-                # spark would throw error when table doesn't exist, where other
-                # CDW would just return and empty list, normalizing the behavior here
-                errmsg = getattr(e, "msg", "")
-                if "Table or view not found" in errmsg or "NoSuchTableException" in errmsg:
-                    pass
-                else:
-                    raise e
+
+        try:
+            rows: List[agate.Row] = self.execute_macro(
+                GET_COLUMNS_IN_RELATION_RAW_MACRO_NAME, kwargs={"relation": relation}
+            )
```

Comment:
This is the `describe table extended` query that Spark uses to return detailed information, including the columns in the table